devinfo/src/block_device.rs

```diff
@@ -0,0 +1,114 @@
+use std::{convert::TryFrom, iter::Iterator};
+use udev::Enumerator;
+use url::Url;
+use uuid::Uuid;
+
+use crate::{
+    DevInfoError,
+    DevInfoError::{NqnInvalid, Udev},
+};
+
+pub(crate) type Failable<T, E = DevInfoError> = std::result::Result<T, E>;
+
+#[derive(Debug)]
+pub enum BlkDev {
+    Scsi(Uuid),
+    Nvmf(Uuid),
+    File(String),
+}
+
+impl TryFrom<&str> for BlkDev {
+    type Error = DevInfoError;
+    /// convert a URI in the form of
+    ///
+    /// scheme:://hostname/first_segment/second/segment
+    ///
+    /// When parsing the scheme iscsi:// or nvmf:// on the first segment is
+    /// considered when dealing with the file:// scheme, we dont do much but concatenate the two segments
+    fn try_from(uri: &str) -> Failable<Self> {
+        let value = Url::parse(uri).map_err(|e| DevInfoError::ParseError {
+            value: e.to_string(),
+        })?;
+
+        let mut nq = value.path_segments().ok_or_else(|| {
+            DevInfoError::NotSupported {
+                value: "The uri contains contains no path segments".to_string(),
+            }
+        })?;
+
+        // for files we dont have to do much, simply merge the first two
+        // elements of the path segment
+        if value.scheme() == "file" {
+            return Ok(BlkDev::File(format!(
+                "{}/{}",
+                nq.next().unwrap(),
+                nq.next().unwrap()
+            )));
+        }
+
+        // this is not a file scheme so we should have a nqn:uuid type layout
+        // here
+        let nq = nq.next().unwrap().split(':').collect::<Vec<_>>();
+
+        if nq.len() != 2 {
+            return Err(DevInfoError::NqnInvalid {
+                value: "the NQN does not contain the expected separator \':\'"
+                    .to_string(),
+            });
+        }
+
+        let uuid = Uuid::parse_str(nq[1]).map_err(|e| NqnInvalid {
+            value: format!("the UUID is invalid {}", e.to_string()),
+        })?;
+
+        match value.scheme() {
+            "iscsi" => Ok(BlkDev::Scsi(uuid)),
+            "nvmf" => Ok(BlkDev::Nvmf(uuid)),
+            _ => Err(DevInfoError::NotSupported {
+                value: "".to_string(),
```

Comment:
Should we return the scheme that isn't supported here?

Reply:
Thanks! fixed :+1: